Step3 리뷰요청 드립니다.

itdar opened the pull request:
안녕하세요 동철님 주말은 잘 보내셨나요!
이번에는 2단계 때 동철님 리뷰해주셨던 내용이랑 비슷한 내용들이 많아서
문서 참고하면서 수월하게 진행했던 것 같습니다 ㅎㅎ;
리뷰 감사합니다!!

itdar committed:
docs: step3 요구사항 정리

itdar committed:
refactor: 요청데이터 처리 로직 분리 리팩토링

- Request 에서 getParameter 시에, RequestLine 의 쿼리스트링 및 Body 의 쿼리스트링포맷 파라미터 둘 다 찾아서 있는 것 반환하도록 수정 (RequestLine 에서 먼저 찾아지면 Body 까지 확인 안함)
- 요청 테스트리소스파일 픽스처로 추가 (GET, POST)
- 불필요하게 데이터 노출시키는 메서드 제거

itdar committed:
refactor: 응답데이터 처리 리팩토링

- Response 객체 안에서 outputStream 작업 하도록 수정
- Response 의 응답 처리 로직을 forward(자원 uri) , forwardBody(바디 내용), sendRedirect(이동할 자원 uri) 메서드로 변경

itdar committed:
refactor: 다형성을 활용해 요청 URL 에 맞는 컨트롤러 분기처리

- Servlet -> Controller 이름 변경
- Controller 인터페이스에서 디폴트로 service(), 필수구현으로 doGet(), doPost() .... 추가예정

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -24,51 +24,23 @@ public RequestHandler(Socket connectionSocket) {
     public void run() {
```
RequestHandler 코드가 깔끔해졌네요 👍

itdar replied:
옙 감사합니다!!

sah3122 commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -84,4 +109,43 @@ public StatusCode getStatus() {
     }
 
+
+    private void writeResponse() {
```
공통 로직 분리 👍

sah3122 commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -84,4 +109,43 @@ public StatusCode getStatus() {
+        }
+    }
+
+    private void responseStatusLine(DataOutputStream dos, String statusLine) throws IOException {
+        dos.writeBytes(statusLine + " \r\n");
+        logger.debug(statusLine);
+    }
+
+    private void responseHeader(DataOutputStream dos, HttpResponseHeader responseHeader) throws IOException {
+        for (String key : responseHeader.keySet()) {
+            dos.writeBytes(key + ": " + responseHeader.getHeader(key) + "\r\n");
+            logger.debug(key + ": " + responseHeader.getHeader(key));
+        }
+        dos.writeBytes("\r\n");
+    }
+
+    private void responseBody(DataOutputStream dos, byte[] body) {
+        try {
+            dos.write(body, 0, body.length);
+            dos.flush();
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+        }
+    }
```
프레임워크를 만드는 입장과 사용하는 입장에서 커스텀할수 있는 api 를 제공하는것도 고민해보시면 좋을것 같아요. 
가령 Response를 제공하는 순간에 특정 Body 를 추가한다던지 그러한 통로를 열어주는것도 중요하다고 생각해요. ㅎㅎ

itdar replied:
아.. 계속 최대한 외부에 노출이 안되도록하고 예상 사용법을 벗어나지 않게 하는 방향으로 생각하면서 만들었던 것 같네요..ㅎㅎ;
말씀하신대로 커스텀이 가능하도록 수정을 해보겠습니다!~

itdar replied:
말씀해주신 내용을 보고 처음 생각으로는 StatusLine / Header / Body 를 각각 값을 넣으면서 바로 응답을 출력하는 메서드를 오픈하는게 좋으려나 했었는데..
 각 파트를 하나씩 커스텀 하는 것은 의미가 없을 것 같아서 결국 아래와 같이 응답 묶음 메서드에서 바디를 넣어주거나.. 하는게 나을 것 같았는데요 ㅎㅎ; 어떻게 생각하시나요? @sah3122 

 ```java
    public void forwardBody(String page) {
        byte[] body = page.getBytes(StandardCharsets.UTF_8);
        forwardBody(body);
    }

    private void forwardBody(byte[] body) {
        this.statusLine = new StatusLine(this.statusLine.getProtocol(), HttpStatus.OK);
        this.body = HttpResponseBody.of(body);

        writeResponse();
    }

...

    public void notFound(String body) {
        this.statusLine = new StatusLine(this.statusLine.getProtocol(), HttpStatus.NOT_FOUND);
        this.body = HttpResponseBody.of(body.getBytes(StandardCharsets.UTF_8));

        writeResponse();
    }
```

sah3122 replied:
public 으로 열어주는것도 좋은 방법이 될것 같아요 😄 
Custom Response 를 만들수 있도록 Response Interface 를 미리 분리해두는것도 좋은 선택지가 될수 있을것 같고요 ㅎㅎ

itdar replied:
아.. 마지막 단계에서 Response 인터페이스를 분리해봐야겠습니다!!

sah3122 commented on src/main/java/webserver/servlet/Controller.java:
```diff
@@ -0,0 +1,22 @@
+        doGet(httpRequest, httpResponse);
+    }
+
+    void doPost(HttpRequest httpRequest, HttpResponse httpResponse);
+
+    void doGet(HttpRequest httpRequest, HttpResponse httpResponse);
```
default 구현을 제공하는것이 어떨까요 ?

```suggestion
    default void doPost(HttpRequest httpRequest, HttpResponse httpResponse) {
       // throw Exception 
    }

    default void doGet(HttpRequest httpRequest, HttpResponse httpResponse) {
       // throw Exception 
    }
```

itdar replied:
옙옙 다같이 디폴트로 만들까하다가 말았었는데 수정해보겠습니다!!

sah3122 commented on src/main/java/webserver/servlet/UserLoginController.java:
```diff
@@ -9,25 +9,30 @@
         if (userFound.fitPassword(password)) {
-            httpResponse.found("http://localhost:8080/index.html");
             httpResponse.addHeader(HttpHeader.SET_COOKIE, "logined=true; Path=/");
+            httpResponse.sendRedirect("http://localhost:8080/index.html");
```
path만 제공하는것과 host를 함께 제공하는것에 대한 차이점을 고민해보세요 😄

itdar replied:
음.. 프로토콜이랑 호스트, 포트를 명시해서 반환하면..
클라이언트 쪽에서는 path 만 다시 파싱해야 다시 사용해야 할 것 같네요 ㅎㅎ;
서버는 프로토콜이나 포트, 도메인이 바뀌면 다같이 수정을 해줘야 할 것 같구요!
path 만 반환하게 수정해보겠습니다!!

sah3122 commented on src/main/java/webserver/servlet/Controller.java:
```diff
@@ -0,0 +1,22 @@
+
+public interface Controller {
+
+    default void service(HttpRequest httpRequest, HttpResponse httpResponse) {
+        if (httpRequest.getMethod() == HttpMethod.POST) {
+            doPost(httpRequest, httpResponse);
+            return;
+        }
+
+        doGet(httpRequest, httpResponse);
+    }
```
PUT, DELETE등과 같이 아직 지원하지 않는 요청이 들어온다면 어떤 응답을 제공해주어야 할지 고민해보세요.

itdar replied:
아.. 405 status 를 여기서 써야할 것 같습니다!
보다보니 추가로 회원가입 post 는 201 로 바꾸어야 할 것 같았는데 302 Found 리다이렉션이네요 ㅎㅎ;

sah3122 reviewed (changes requested):
3단계 리팩토링 잘해주셨네요 👍 
몇가지 고민거리를 남겨두었는데 확인 부탁드려요 🙏

itdar committed:
refactor: 동철님 리뷰사항 수정 리팩토링

- 리다이렉트 uri 에서 path 만 전달하도록 수정
- Controller interface 에서 get/post 등 디폴트메서드로 변경
- StatusCode -> HttpStatus enum class 이름 변경
- http 미허용 메서드 응답 추가

itdar committed:
chore: code reformatting 및 import optimizing (IDE 제공기능)

sah3122 commented on src/main/java/webserver/response/HttpResponse.java:
```diff
@@ -62,29 +63,42 @@ public void forwardBody(String page) {
     }
 
+    public void notFound(String body) {
+        this.statusLine = new StatusLine(this.statusLine.getProtocol(), HttpStatus.NOT_FOUND);
+        this.body = HttpResponseBody.of(body.getBytes(StandardCharsets.UTF_8));
+
+        writeResponse();
+    }
+
+    public void methodNotAllowed() {
+        this.statusLine = new StatusLine(this.statusLine.getProtocol(), HttpStatus.METHOD_NOT_ALLOWED);
+
+        writeResponse();
+    }
```
예외 처리 👍

sah3122 commented on src/main/java/webserver/servlet/Controller.java:
```diff
@@ -12,11 +12,20 @@ default void service(HttpRequest httpRequest, HttpResponse httpResponse) {
 
-    void doPost(HttpRequest httpRequest, HttpResponse httpResponse);
+    default void doPost(HttpRequest httpRequest, HttpResponse httpResponse) {
+        throw new UnsupportedOperationException();
+    }
 
-    void doGet(HttpRequest httpRequest, HttpResponse httpResponse);
+    default void doGet(HttpRequest httpRequest, HttpResponse httpResponse) {
+        throw new UnsupportedOperationException();
+    }
```
default 구현 👍 

Servlet 인터페이스와 HttpServlet 인터페이스가 어떻게 분리되어있는지, 구현해주신 결과와 어떤 차이점이 있는지 한번 보시면 좋을것 같아 링크 남겨드려요.

https://docs.oracle.com/javaee/7/api/javax/servlet/Servlet.html
https://docs.oracle.com/javaee/7/api/javax/servlet/http/HttpServlet.html

개인적으로 스프링 프레임워크 기준 DispatherServlet, FrameworkServlet 의 service 구현부를 파보는것이 도움이 많이 되었습니다 😄

itdar replied:
오 링크 감사합니다 봐보겠습니다!!

sah3122 commented on src/main/java/webserver/servlet/UserCreateController.java:
```diff
@@ -17,12 +17,7 @@ public void doPost(HttpRequest httpRequest, HttpResponse httpResponse) {
-    @Override
-    public void doGet(HttpRequest httpRequest, HttpResponse httpResponse) {
-
+        httpResponse.sendRedirect("/index.html");
```
피드백 반영 👍 
redirect 상태값도 다양하게 있으니 아래 문서를 참고해보세요 ㅎㅎ 
리뷰를 통해 저도 많이 배우고 있습니다 🙏 

https://developer.mozilla.org/ko/docs/Web/HTTP/Redirections

itdar replied:
아.. 그러네요 리다이렉트도 상태코드를 같이 넣어주는게 좋겠네요 ㅎㅎ; 감사합니다!!

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 💯 
개인적으로 첫번째 미션을 진행할때 2,3 단계가 가장 힘들었고 그만큼 기억에 많이 남는 미션이었던것 같아요. 
이번 미션을 통해 학습하신것들을 한번 되돌아 보며 형진님의 것으로 만들어보세요 👍
